polyfills/systemxhr/systemxhr.js

```diff
@@ -0,0 +1,430 @@
+//  observers!)
+
+// createXMLHttpRequest =>
+//   * Request: { id: requestId,
+//                data: {
+//                       operation: 'createXMLHttpRequest',
+//                       options: options
+//                      }
+//              }
+//   * Answer:  { id: requestId, error: error}
+//     (answer.error truthy means there was an error)
+// Once you get an answer you can assume than the id is the identifier of the
+// remote XMLHttpRequest
+
+// For all the operations over a XMLHttpRequest:
+//  * Request: { id: requestId,
+//               data: {
+//                 xhrId: xhrId,
+//                 operation: methodName,
+//                 params: params
+//               }
+//             }
+// * Answer (only those methods that return something): { id: requestId,
+//             result|error: Whatever }
+
+// For the EventTarget operations
+//  Request: { id: requestId,
+//             data: {
+//             operation: 'addEventListener|removeEventListener|dispatchEvent',
+//             type: eventType (only addEventListener and removeEventListener),
+//             useCapture: true|false (only addEventListener),
+//             event: eventToDispatch (only dispatchEvent)
+//           },
+// Answer: When invoked:
+//    { id: requestId,
+//      data: EventTargetEvent }
+
+
+(function(window) {
+
+  'use strict';
+
+  function debug(text) {
+    console.log('*-*-*- System XHR PF: ' + text);
+  }
+
+  // Wishful thinking at the moment...
+  const SYSTEMXHR_SERVICE = 'https://systemxhrservice.gaiamobile.org';
+
+  // XMLHttpRequest polyfill..
+  function FakeXMLHttpRequest(reqId, extraData) {
+
+    var _resolve, _reject;
+    var _systemxhr = new Promise((resolve, reject) => {
+      _resolve = resolve;
+      _reject = reject;
+    });
+
+    // This will hold the remote object, once the lock is actually created..
+    var _xhrId = null;
+    _systemxhr.then(id => _xhrId = id);
+    var self = this;
+
+    function OnReadyStateChangeRequest(reqId, extraData) {
+      this.serialize = () => {
+        return {
+          id: reqId,
+          data: extraData,
+          processAnswer: answer => {
+            if (answer.event) {
+              self._updateXMLHttpRequestObject(answer.event);
+              self.onreadystatechange && self.onreadystatechange(answer.event);
+            }
+          }
+        };
+      };
+    }
+
+    function OnChangeRequest(reqId, extraData) {
+      this.serialize = () => {
+        return {
+          id: reqId,
+          data: extraData,
+          processAnswer: answer => {
+            if (answer.event) {
+              self['_' + extraData.operation](answer.event);
+            }
+          }
+        };
+      };
+    }
+
+
+    function _listenerCallback(evt, cb) {
+      this._updateXMLHttpRequestObject.call(this, evt);
+
+      cb(evt);
+    }
+
+    FakeEventTarget.call(_listenerCallback.bind(this), _systemxhr);
+
+    this.abort = function() {
```

Comment:
Those three functions can be folded into something like:
```
[{method: 'abort', numParams: 0}, {method: 'open', numParams: 5}, {method: 'overrideMimeType', numParams: 1}].forEach(info => 
  this[info.method] = navConnPromise.methodCall.bind(this, {methodName: info.method, numParams: info.numParams, returnValue: VoidRequest, promise: _systemxhr, field: 'xhrId'}));
```
(prettified a little :P)

})